src/main/scala/logger/SymbExLogger.scala

```diff
@@ -192,36 +185,42 @@ import scala.util.{Failure, Success, Try}
   * // SymbExLogger.currentLog().markReachable(uidBranchPoint)
   */
 
-object SymbExLogger {
-  /** Collection of logged Method/Predicates/Functions. **/
-  var memberList: Seq[SymbLog] = Seq[SymbLog]()
-  private var uidCounter = 0
-
-  var enabled = false
-  var logConfig: LogConfig = LogConfig.default()
-
-  def freshUid(): Int = {
-    val uid = uidCounter
-    uidCounter = uidCounter + 1
-    uid
+case object SymbExLogger {
+  def ofConfig(c: Config): SymbExLogger[_ <: MemberSymbExLogger] = {
+    if(c.ideModeAdvanced())
+      SymbExLog(parseLogConfig(c))
+    else
+      NoopSymbExLog
   }
 
-  def getRecordConfig(d: DataRecord): Option[RecordConfig] = {
-    for (rc <- logConfig.recordConfigs) {
-      if (rc.kind.equals(d.toTypeString)) {
-        rc.value match {
-          case Some(value) => if (value.equals(d.toSimpleString)) return Some(rc)
-          case _ => return Some(rc)
-        }
-      }
+  private def parseLogConfig(c: Config): LogConfig = {
+    var logConfigPath = Try(c.logConfig())
+    logConfigPath = logConfigPath.filter(path => Files.exists(Paths.get(path)))
+    val source = logConfigPath.map(path => scala.io.Source.fromFile(path))
+    val fileContent = source.map(s => s.getLines().mkString)
+    val jsonAst = fileContent.flatMap(content => Try(content.parseJson))
+    val logConfig = jsonAst.flatMap(ast => Try(ast.convertTo[LogConfig]))
+    logConfig match {
+      case Success(convertedConfig) => convertedConfig
+      case Failure(_) => LogConfig.default()
     }
-    None
   }
+}
 
-  /**
-    * stores the last SMT solver statistics to calculate the diff
-    */
-  private var prevSmtStatistics: Map[String, String] = new Map()
+abstract class SymbExLogger[Log <: MemberSymbExLogger]() {
+  /** Collection of logged Method/Predicates/Functions. */
+  protected var members: immutable.Map[ast.Member, Log] = Map.empty
+  protected var closed: Boolean = false
+  def isClosed: Boolean = closed
+  protected val uidCounter: AtomicInteger = new AtomicInteger(0)
+
+  protected def newEntityLogger(member: ast.Member, pcs: PathConditionStack): Log
+
+  def freshUid(): Int = uidCounter.getAndIncrement()
+
+  def whenEnabled(f: => Unit): Unit = f
```

Comment:
Would anything change if we made the closure explicit: `f: () => Unit` and `... = f()`? I find the current Scala syntax very unintuitive, in particular for people that don't know Scala very well.

Just `f()` would already be more explicit.

Reply:
Yes, it would change at the usage site, e.g.:

```
def defn(f: => Unit): Unit = f

def use(): Unit = {
  defn {
    println()
  }
}
```

becomes e.g.:

```
def defn(f: () => Unit): Unit = f()

def use(): Unit = {
  defn { () =>
    println()
  }
}
```

I agree that it's a fairly arcane scala feature. I sometimes use it when the usage of it is intuitive (like I think it is here), at the cost of a weird definition. Happy to make it more explicit, though.

Reply:
Thanks, I retract my request. The syntax overhead at call-site is not worth the additional explicitness at definition-site.